### ISSUE TITLE
Emission shape ring for CPUParticles2D

### ISSUE THREAD

**Issue by StamesJames** [godotengine/godot-proposals#10303]
### Describe the project you are working on

A 2D Shoot'em up game

### Describe the problem or limitation you are having in your project

I am trying to visualize a laser loading by pulling in particles in some radius. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

With the emission shape ring, particles would be generated in a ring.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Add a enum Variant EMISSION_SHAPE_RING to the EmissionShape Enum. Give this emission shape the properties emission_ring_inner_radius and emission_ring_outer_radius and uniformly sample in the ring between those circles the following way:

```c++
case EMISSION_SHAPE_RING: {
real_t t = MATH_TAU * Math::randf();
real_t radius = emission_ring_inner_radius + Math:sqrt(Math::randf()) * (emission_ring_outer_radius - emission_ring_inner_radius);
p.transform[2] = Vector2(Math::cos(t), Math::in(t)) * radius;
}
```

I think I should be able to write a PR for this.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I don't think it can be worked around with a few lines of code but maybe with emission masks by loading in a texture of a ring. But this would not be very flexible because you wouldn't be able to change the sizes of the rings easily.

### Is there a reason why this should be core and not an add-on in the asset library?

Because there already exists a similar feature in 3D and also I think there was such an emission shape in the 3.x versions of the particle systems (I have seen it in a tutorial once)

**Comment by Calinou** [comment]
Feel free to open a pull request for this :slightly_smiling_face: 

> Because there already exists a similar feature in 3D and also I think there was such an emission shape in the 3.x versions of the particle systems (I have seen it in a tutorial once)

This emission shape was never implemented in 2D in the first place, only in 3D:

![image](https://github.com/user-attachments/assets/02bb47c5-854e-447c-82de-b35023919557)

**Comment by StamesJames** [comment]
Ok nice I will look into this 😀 this will be my first Godot PR 😁

**Comment by QbieShay** [comment]
I am late to this proposal, i wonder why not adding an inner radius to the circle emitter?

**Comment by StamesJames** [comment]
@QbieShay 
The current circle emitter samples the points in the following way
```c++
case EMISSION_SHAPE_SPHERE: {
	real_t t = Math_TAU * Math::randf();
	real_t radius = emission_sphere_radius * Math::randf();
	p.transform[2] = Vector2(Math::cos(t), Math::sin(t)) * radius;
} break;
```
The particles are not sampled uniformly in the circle because there are more particles in the middle of the circle. 
I wrote the ring emission shape so that the particles are sampled uniformly, and I did not want to change the particle distribution of the circle.

**Comment by StamesJames** [comment]
Oh and also this way it is more similar to the 3D particles and it directly works with GPUParticles because 2D and 3D Particles share the same ParticleProcessMaterial.

**Comment by QbieShay** [comment]
iirc the gpu ones were adapted to have a uniform distribution. at least from what i remember. if they do, then they should all just have the same distribution.

**Comment by StamesJames** [comment]
I don't think so. In 2D when I create a CPUParaticles2D with the sphere shape and convert it to a GPUParticles2D I get this:
![Image](https://github.com/user-attachments/assets/06335ee8-62ba-4767-af20-fbb8552fb252)

With the Ring Shape and an inner radius of 0, I get this:
![Image](https://github.com/user-attachments/assets/785bf446-65c1-4708-add5-c84b2ddc17a9)

In the Sphere shape, there are a bit more particles in the middle.
